homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,389 @@
+    # tracked tasks (hass.loop.create_task). This way HA won't stall startup
+    # forever until a connection is successful.
+
+    async def complete_setup() -> None:
+        """Complete the config entry setup."""
+        tasks = []
+        for component in HA_COMPONENTS:
+            tasks.append(hass.config_entries.async_forward_entry_setup(
+                entry, component))
+        await asyncio.wait(tasks)
+
+        # If first connect fails, the next re-connect will be scheduled
+        # outside of _pending_task, in order not to delay HA startup
+        # indefinitely
+        await try_connect(is_disconnect=False)
+
+    hass.async_create_task(complete_setup())
+    return True
+
+
+async def _setup_auto_reconnect_logic(hass: HomeAssistantType,
+                                      cli: 'APIClient',
+                                      entry: ConfigEntry, host: str):
+    """Set up the re-connect logic for the API client."""
+    from aioesphomeapi.client import APIConnectionError
+
+    async def try_connect(tries: int = 0, is_disconnect: bool = True) -> None:
+        """Try connecting to the API client. Will retry if not successful."""
+        if entry.entry_id not in hass.data[DOMAIN]:
+            # When removing/disconnecting manually
+            return
+
+        data = hass.data[DOMAIN][entry.entry_id]  # type: RuntimeEntryData
+        data.available = False
+        data.async_update_device_state(hass)
+
+        if tries != 0:
+            # If not first re-try, wait and print message
+            wait_time = min(2**tries, 300)
+            _LOGGER.info("Trying to reconnect in %s seconds", wait_time)
+            await asyncio.sleep(wait_time)
+
+        if is_disconnect and tries == 0:
+            # This can happen often depending on WiFi signal strength.
+            # So therefore all these connection warnings are logged
+            # as infos. The "unavailable" logic will still trigger so the
+            # user knows if the device is not connected.
+            _LOGGER.info("Disconnected from API")
+
+        try:
+            await cli.connect()
+            await cli.login()
+        except APIConnectionError as error:
+            _LOGGER.info("Can't connect to esphomelib API for '%s' (%s)",
+                         host, error)
+            # Schedule re-connect in event loop in order not to delay HA
+            # startup. First connect is scheduled in tracked tasks.
+            data.reconnect_task = \
+                hass.loop.create_task(try_connect(tries + 1, is_disconnect))
+        else:
+            _LOGGER.info("Successfully connected to %s", host)
+
+    cli.on_disconnect = try_connect
+    return try_connect
+
+
+async def _cleanup_instance(hass: HomeAssistantType,
+                            entry: ConfigEntry) -> None:
+    """Cleanup the esphomelib client if it exists."""
+    data = hass.data[DOMAIN].pop(entry.entry_id)  # type: RuntimeEntryData
```

Comment:
Make sure to unsubscribe any remaining component wide listeners here, eg the on home assistant stop event listener.

Reply:
What will happen if we remove the config entry, but don't unsubscribe the home assistant stop event listener? Will there be a KeyError here?

Reply:
Oh sorry, I was looking for uses of `async_dispatcher_connect` and forgot that callback